sdk/src/sharedTest/java/io/customer/sdk/CustomerIOClientTest.kt

```diff
@@ -253,6 +253,44 @@ class CustomerIOClientTest : BaseTest() {
         prefRepository.getDeviceToken() shouldBeEqualTo givenDeviceToken
     }
 
+    // addCustomDeviceAttributes
+
+    @Test
+    fun addCustomDeviceAttributes_givenNoPushToken_expectDoNotRegisterPushToken() {
+        val givenAttributes = mapOf(String.random to String.random)
+
+        customerIOClient.addCustomDeviceAttributes(givenAttributes)
+
+        // no token registered
+        verifyNoInteractions(backgroundQueueMock)
+    }
+
+    @Test
+    fun addCustomDeviceAttributes_givenExistingPushToken_expectRegisterPushTokenAndAttributes() {
+        val givenAttributes = mapOf(String.random to String.random)
+        val givenDeviceToken = String.random
+        val givenIdentifier = String.random
+        prefRepository.saveDeviceToken(givenDeviceToken)
+        prefRepository.saveIdentifier(givenIdentifier)
+
+        customerIOClient.addCustomDeviceAttributes(givenAttributes)
+
+        // a token got registered
+        verify(backgroundQueueMock).addTask(
+            QueueTaskType.RegisterDeviceToken,
+            RegisterPushNotificationQueueTaskData(
+                givenIdentifier,
+                Device(
+                    token = givenDeviceToken,
+                    lastUsed = dateUtilStub.givenDate,
+                    attributes = deviceStore.buildDeviceAttributes() + givenAttributes
+                )
+            ),
+            groupStart = QueueTaskGroup.RegisterPushToken(givenDeviceToken),
+            blockingGroups = listOf(QueueTaskGroup.IdentifyProfile(givenIdentifier))
+        )
```

Comment:
I do not like the code used here to test if `identify()` was called. This will be addressed with [a refactor splitting up the `CustomerIOClient` class](https://github.com/customerio/issues/issues/7160). These lines of code could then be much easier to understand and easier to maintain. I imagine the code will be changed to something similar to this:

```kotlin
verify(deviceRepository).registerDeviceToken(givenDeviceToken)
```

Reply:
implemented https://github.com/customerio/customerio-android/pull/103